### DISCUSSION TITLE
Design suggestion for recipe functions

### DISCUSSION THREAD

**Discussion by fpavogt** [MeteoSwiss/dvas#149] [Ideas]
Hey @modolol, 

I've given some more thoughts about how to handle dvas recipe functions in the cleanest way possible. The two challenges we face is the need to deal with many variables (either one after the other, or all at once), and many flights (either one after the other, or all at once). 

On the one hand, it would be great if the core step function would not need too much about all the mechanics handling them. On the other hand, the idea of have a `RecipeSteps` class as generic as possible would be ideal, while keep the `uaii22.rcp` yaml file as simple as possible. 

I've been toying around the idea of using decorators to let function designers specify if said function should be sequential or not. Something along the lines:

```

from .dynamic import ALL_FLIGHTS; THIS_FLIGHT

def for_all_flights(func):
    """ This is a decorator"""
    def wrapper(**kwargs):
        """ This is where we loop around the original function """

        for flight in ALL_FLIGHTS:
            THIS_FLIGHT = flight
            func(**kwargs)
 
    return wrapper
```

And then, the function defined in another file, would be:
```
from ...recipe import for_all_flights
from ...dynamic import THIS_FLIGHT


@for_all_flights
def dummy_step_fct(args_1=14):

    """ This is what the user would code. He/She would have access to the global variable THIS_FLIGHT which would get automatically adjusted by the decorator for each step."""

   print(args_1, THIS_FLIGHT)
```

**Problem:** I cannot get THIS_FLIGHT and ALL_FLIIGHTS to be set/read dynamically accross the `dvas_recipe` package. Basically, what I'd need is a way to set/read them on the go. I could do it by writing them to file ... but could we do something more elegant ? Do you have any idea why the above does not work ? Is it because I do relative imports rather than absolute ones ?

**Comment by fpavogt** [comment]
Big discovery of the day: the code above works if I do:
```
from . import dynamic # instead of from .dynamic import ALL_FLIGHTS
```
With this, I should be able to get it working in the way I hoped. It should be clean from every perspective thanks to these decorators.

**Reply by modolol** [reply]
I'm not sure if it would be working. But I say this without being sure to grasp the full difficulty of the problem.
What do you think about using the Singleton pattern 🤣 (`dvas.helper.SingleInstanceMetaClass`) as used for global variables in dvas like `dvas.environ.GlobalPathVariablesManager`

**Reply by fpavogt** [reply]
I think it is applicable. I saw people referring to this pattern when talking about the solution above. It seems that creating a module dedicated to storing global variable is sort of like a Singleton pattern. Let me clean things up to get something that runs, and we take a closer look.

**Comment by fpavogt** [comment]
Locking this thread for now. The ideas above were eventually implemented in #158.
